// @rmwc/notched-outline/dist/foundation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = require("react");
var base_1 = require("@rmwc/base");
var notched_outline_1 = require("@material/notched-outline");
exports.useNotchedOutlineFoundation = function (props) {
    var _a = base_1.useFoundation({
        props: props,
        elements: {
            rootEl: true,
            notchedEl: true
        },
        foundation: function (_a) {
            var rootEl = _a.rootEl, notchedEl = _a.notchedEl;
            return new notched_outline_1.MDCNotchedOutlineFoundation({
                addClass: function (className) { return rootEl.addClass(className); },
                removeClass: function (className) { return rootEl.removeClass(className); },
                setNotchWidthProperty: function (width) {
                    return notchedEl.setStyle('width', width + 'px');
                },
                removeNotchWidthProperty: function () { return notchedEl.setStyle('width', ''); }
            });
        }
    }), foundation = _a.foundation, elements = tslib_1.__rest(_a, ["foundation"]);
    var rootEl = elements.rootEl;
    var labelRef = react_1.useRef();
    react_1.useEffect(function () {
        !!props.notch ? foundation.notch(props.notch) : foundation.closeNotch();
    }, [props.notch, foundation]);
    react_1.useEffect(function () {
        var _a;
        labelRef.current =
            ((_a = rootEl.ref) === null || _a === void 0 ? void 0 : _a.querySelector('.mdc-floating-label')) || undefined;
        var label = labelRef.current;
        if (label) {
            label.style.transitionDuration = '0s';
            rootEl.addClass(notched_outline_1.MDCNotchedOutlineFoundation.cssClasses.OUTLINE_UPGRADED);
            requestAnimationFrame(function () {
                label && (label.style.transitionDuration = '');
            });
        }
        else {
            rootEl.addClass(notched_outline_1.MDCNotchedOutlineFoundation.cssClasses.NO_LABEL);
        }
    }, [rootEl]);
    return tslib_1.__assign({ foundation: foundation }, elements);
};
