// @rmwc/toggleable/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var base_1 = require("@rmwc/base");
var formfield_1 = require("@rmwc/formfield");
function useToggleFoundation(props) {
    var className = props.className, style = props.style, rootProps = props.rootProps, label = props.label, children = props.children, disabled = props.disabled;
    var hasLabel = props.label || props.children;
    var id = base_1.useId('toggle-', props);
    var renderToggle = function (toggle) {
        /**
         * We have to conditionally wrap our checkbox in a formfield
         * If we have a label
         */
        if (hasLabel) {
            return (react_1.default.createElement(formfield_1.FormField, tslib_1.__assign({}, rootProps, { className: className, style: style }),
                toggle,
                react_1.default.createElement("label", { id: id + 'label', htmlFor: id },
                    label,
                    children)));
        }
        else {
            return toggle;
        }
    };
    var toggleRootProps = hasLabel
        ? { disabled: disabled }
        : tslib_1.__assign({ className: className,
            style: style,
            disabled: disabled }, rootProps);
    return {
        id: id,
        renderToggle: renderToggle,
        toggleRootProps: toggleRootProps
    };
}
exports.useToggleFoundation = useToggleFoundation;
