// @rmwc/typography/dist/readme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.galleryExample = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var _doc_utils_1 = require("@doc-utils");
var generated_props_json_1 = tslib_1.__importDefault(require("./generated-props.json"));
var generated_examples_json_1 = tslib_1.__importDefault(require("./generated-examples.json"));
var _1 = require(".");
var provider_1 = require("../provider");
function default_1() {
    return (react_1.default.createElement(_doc_utils_1.Docs, { title: "Typography", lead: "Material Design\u2019s text sizes and styles were developed to balance content density and reading comfort under typical usage conditions.", module: "@rmwc/typography", styles: ['@material/typography/dist/mdc.typography.css'], docsLink: "https://material.io/develop/web/components/typography/", examples: generated_examples_json_1.default },
        react_1.default.createElement(_doc_utils_1.DocsSubtitle, null, "Basic Usage"),
        react_1.default.createElement(_doc_utils_1.DocsExample, null,
            react_1.default.createElement(react_1.default.Fragment, null,
                react_1.default.createElement(_1.Typography, { use: "headline1" }, "headline1"),
                react_1.default.createElement(_1.Typography, { use: "headline2" }, "headline2"),
                react_1.default.createElement(_1.Typography, { use: "headline3" }, "headline3"),
                react_1.default.createElement(_1.Typography, { use: "headline4" }, "headline4"),
                react_1.default.createElement(_1.Typography, { use: "headline5" }, "headline5"),
                react_1.default.createElement(_1.Typography, { use: "headline6" }, "headline6"),
                react_1.default.createElement(_1.Typography, { use: "subtitle1" }, "subtitle1"),
                react_1.default.createElement(_1.Typography, { use: "subtitle2" }, "subtitle2"),
                react_1.default.createElement(_1.Typography, { use: "body1" }, "body1"),
                react_1.default.createElement(_1.Typography, { use: "body2" }, "body2"),
                react_1.default.createElement(_1.Typography, { use: "caption" }, "caption"),
                react_1.default.createElement(_1.Typography, { use: "button" }, "button"),
                react_1.default.createElement(_1.Typography, { use: "overline" }, "overline"))),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Specify tag" },
            react_1.default.createElement(_1.Typography, { use: "headline6", tag: "h3" }, "headline6")),
        react_1.default.createElement(_doc_utils_1.DocsSubtitle, null, "Provider Setup"),
        react_1.default.createElement(_doc_utils_1.DocsP, null, "You can globally map specify type classes to HTML tags for semantics, or specify a default to be used instead of `span`."),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Specify tag" },
            react_1.default.createElement(provider_1.RMWCProvider, { typography: {
                    /** Make all Typography components default to <div>  */
                    defaultTag: 'div',
                    /** Make headline5 <h5>  */
                    headline5: 'h5',
                    /** Make body2 <p>  */
                    body2: 'p',
                    /** Use your own component  */
                    body1: function (_a) {
                        var children = _a.children, rest = tslib_1.__rest(_a, ["children"]);
                        return (react_1.default.createElement("span", null,
                            react_1.default.createElement("b", null,
                                children,
                                "!!!")));
                    }
                } },
                react_1.default.createElement(_1.Typography, { use: "headline6" }, "Rendered default `div`"),
                react_1.default.createElement(_1.Typography, { use: "headline5" }, "Rendered with `h5`"),
                react_1.default.createElement(_1.Typography, { use: "body2" }, "Rendered with `p`"),
                react_1.default.createElement(_1.Typography, { use: "body1" }, "Custom rendering"))),
        react_1.default.createElement(_doc_utils_1.DocProps, { src: generated_props_json_1.default, components: [{ displayName: 'Typography', component: _1.Typography }] })));
}
exports.default = default_1;
exports.galleryExample = (react_1.default.createElement("div", { style: { transform: 'scale(1.5)' } },
    react_1.default.createElement(_1.Typography, { tag: "div", use: "headline6" }, "headline6"),
    react_1.default.createElement(_1.Typography, { tag: "div", use: "body1" }, "body1"),
    react_1.default.createElement(_1.Typography, { tag: "div", use: "caption" }, "caption")));
