// @rmwc/checkbox/dist/foundation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useCheckboxFoundation = void 0;
var tslib_1 = require("tslib");
var toggleable_1 = require("@rmwc/toggleable");
var base_1 = require("@rmwc/base");
var checkbox_1 = require("@material/checkbox");
var react_1 = require("react");
exports.useCheckboxFoundation = function (props) {
    var _a = toggleable_1.useToggleFoundation(props), renderToggle = _a.renderToggle, toggleRootProps = _a.toggleRootProps, id = _a.id;
    var _b = base_1.useFoundation({
        props: props,
        elements: {
            rootEl: true,
            checkboxEl: true
        },
        foundation: function (_a) {
            var rootEl = _a.rootEl, checkboxEl = _a.checkboxEl, getProps = _a.getProps;
            return new checkbox_1.MDCCheckboxFoundation({
                addClass: function (className) { return rootEl.addClass(className); },
                removeClass: function (className) { return rootEl.removeClass(className); },
                setNativeControlAttr: function (attr, value) {
                    return checkboxEl.setProp(attr, value);
                },
                removeNativeControlAttr: function (attr) {
                    return checkboxEl.removeProp(attr);
                },
                isIndeterminate: function () { return !!getProps().indeterminate; },
                isChecked: function () {
                    var _a;
                    return getProps().checked !== undefined
                        ? !!getProps().checked
                        : !!((_a = checkboxEl.ref) === null || _a === void 0 ? void 0 : _a.checked);
                },
                hasNativeControl: function () { return !!checkboxEl.ref; },
                setNativeControlDisabled: function (disabled) {
                    return checkboxEl.setProp('disabled', disabled);
                },
                forceLayout: function () { var _a; return (_a = rootEl.ref) === null || _a === void 0 ? void 0 : _a.offsetWidth; },
                isAttachedToDOM: function () { return true; }
            });
        }
    }), foundation = _b.foundation, elements = tslib_1.__rest(_b, ["foundation"]);
    var rootEl = elements.rootEl, checkboxEl = elements.checkboxEl;
    // Handles syncing of indeterminate state
    var doSync = react_1.useCallback(function () {
        if (checkboxEl.ref) {
            checkboxEl.ref.indeterminate = Boolean(props.indeterminate);
        }
        window.requestAnimationFrame(function () {
            foundation.handleChange();
        });
    }, [props.indeterminate, foundation, checkboxEl.ref]);
    react_1.useEffect(function () {
        doSync();
    }, [doSync]);
    // Callback handling
    var handleAnimationEnd = function (evt) {
        var _a;
        (_a = props.onAnimationEnd) === null || _a === void 0 ? void 0 : _a.call(props, evt);
        foundation.handleAnimationEnd();
    };
    var handleOnChange = function (evt) {
        var _a;
        (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, evt);
        doSync();
    };
    rootEl.setProp('onAnimationEnd', handleAnimationEnd, true);
    checkboxEl.setProp('onChange', handleOnChange, true);
    return tslib_1.__assign({ foundation: foundation, renderToggle: renderToggle, toggleRootProps: toggleRootProps, id: id }, elements);
};
