// @rmwc/switch/dist/foundation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useSwitchFoundation = void 0;
var tslib_1 = require("tslib");
var react_1 = require("react");
var toggleable_1 = require("@rmwc/toggleable");
var base_1 = require("@rmwc/base");
var switch_1 = require("@material/switch");
exports.useSwitchFoundation = function (props) {
    var _a = toggleable_1.useToggleFoundation(props), renderToggle = _a.renderToggle, toggleRootProps = _a.toggleRootProps, id = _a.id;
    var _b = base_1.useFoundation({
        props: props,
        elements: {
            rootEl: true,
            checkboxEl: true
        },
        foundation: function (_a) {
            var rootEl = _a.rootEl, checkboxEl = _a.checkboxEl;
            return new switch_1.MDCSwitchFoundation({
                addClass: function (className) { return rootEl.addClass(className); },
                removeClass: function (className) { return rootEl.removeClass(className); },
                setNativeControlChecked: function (checked) {
                    return checkboxEl.setProp('checked', checked);
                },
                setNativeControlDisabled: function (disabled) {
                    return checkboxEl.setProp('disabled', disabled);
                },
                setNativeControlAttr: function (attr, value) {
                    return rootEl.setProp(attr, value);
                }
            });
        }
    }), foundation = _b.foundation, elements = tslib_1.__rest(_b, ["foundation"]);
    var checkboxEl = elements.checkboxEl;
    // On mount, sync the values with the native checkbox
    react_1.useEffect(function () {
        checkboxEl.ref &&
            foundation.updateCheckedStyling_(checkboxEl.ref.checked);
        checkboxEl.ref &&
            foundation.setDisabled(checkboxEl.ref.disabled);
    }, [checkboxEl.ref, foundation]);
    // sync checked
    react_1.useEffect(function () {
        if (props.checked !== undefined) {
            foundation.updateCheckedStyling_(props.checked);
        }
    }, [props.checked, foundation]);
    // sync disabled
    react_1.useEffect(function () {
        if (props.disabled !== undefined) {
            foundation.setDisabled(props.disabled);
        }
    }, [props.disabled, foundation]);
    // Callback handling
    var handleOnChange = function (evt) {
        var _a;
        foundation.handleChange(evt);
        (_a = props.onChange) === null || _a === void 0 ? void 0 : _a.call(props, evt);
    };
    checkboxEl.setProp('onChange', handleOnChange, true);
    return tslib_1.__assign({ foundation: foundation, renderToggle: renderToggle, toggleRootProps: toggleRootProps, id: id }, elements);
};
