// @rmwc/switch/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Switch = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var base_1 = require("@rmwc/base");
var ripple_1 = require("@rmwc/ripple");
var foundation_1 = require("./foundation");
/** A Switch component. */
exports.Switch = base_1.createComponent(function Switch(props, ref) {
    var _a;
    var _b = foundation_1.useSwitchFoundation(props), renderToggle = _b.renderToggle, id = _b.id, toggleRootProps = _b.toggleRootProps, rootEl = _b.rootEl, checkboxEl = _b.checkboxEl;
    var rootClassName = base_1.useClassNames(toggleRootProps, ['mdc-switch']);
    var children = props.children, className = props.className, label = props.label, style = props.style, inputRef = props.inputRef, foundationRef = props.foundationRef, rest = tslib_1.__rest(props, ["children", "className", "label", "style", "inputRef", "foundationRef"]);
    var renderedSwitch = (react_1.default.createElement(base_1.Tag, tslib_1.__assign({}, rootEl.props(tslib_1.__assign(tslib_1.__assign({}, toggleRootProps), { className: rootClassName })), { ref: ref }),
        react_1.default.createElement(SwitchTrack, null),
        react_1.default.createElement(SwitchThumbUnderlay, null,
            react_1.default.createElement(SwitchThumb, null),
            react_1.default.createElement("input", tslib_1.__assign({}, checkboxEl.props(tslib_1.__assign(tslib_1.__assign({}, rest), { className: 'mdc-switch__native-control' })), { type: "checkbox", id: id, ref: base_1.mergeRefs(checkboxEl.setRef, inputRef), role: "switch", "aria-checked": (_a = rest.checked) !== null && _a !== void 0 ? _a : rest['aria-checked'] }))),
        react_1.default.createElement(SwitchKnob, null)));
    return renderToggle(renderedSwitch);
});
/*********************************************************************
 * Bits
 *********************************************************************/
var SwitchTrack = react_1.default.memo(function SwitchTrack() {
    return react_1.default.createElement("div", { className: "mdc-switch__track" });
});
var SwitchKnob = react_1.default.memo(function SwitchKnob() {
    return react_1.default.createElement("div", { className: "mdc-switch__knob" });
});
var SwitchThumb = react_1.default.memo(function SwitchThumb() {
    return react_1.default.createElement("div", { className: "mdc-switch__thumb" });
});
var SwitchThumbUnderlay = ripple_1.withRipple({
    unbounded: true,
    surface: false
})(function SwitchThumbUnderlay(_a) {
    var className = _a.className, rest = tslib_1.__rest(_a, ["className"]);
    return (react_1.default.createElement("div", tslib_1.__assign({ className: base_1.classNames(className, 'mdc-switch__thumb-underlay') }, rest)));
});
