// @rmwc/provider/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RMWCProvider = exports.useProviderContext = exports.ProviderContext = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
// Default provider options
var providerDefaults = {
    ripple: true,
    tooltip: {
        align: 'top',
        showArrow: false,
        activateOn: ['hover', 'focus'],
        enterDelay: 0,
        leaveDelay: 0
    },
    typography: {},
    icon: {
        icon: '',
        basename: 'material-icons',
        prefix: '',
        strategy: 'auto',
        render: undefined
    }
};
exports.ProviderContext = react_1.default.createContext(providerDefaults);
exports.useProviderContext = function () { return react_1.default.useContext(exports.ProviderContext); };
/** A provider for setting global options in RMWC. */
exports.RMWCProvider = function (_a) {
    var children = _a.children, rest = tslib_1.__rest(_a, ["children"]);
    return (react_1.default.createElement(exports.ProviderContext.Provider, { value: tslib_1.__assign(tslib_1.__assign({}, providerDefaults), rest) }, children));
};
