// @rmwc/textfield/dist/readme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.galleryExample = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var _doc_utils_1 = require("@doc-utils");
var generated_props_json_1 = tslib_1.__importDefault(require("./generated-props.json"));
var generated_examples_json_1 = tslib_1.__importDefault(require("./generated-examples.json"));
var _1 = require(".");
function default_1() {
    return (react_1.default.createElement(_doc_utils_1.Docs, { title: "Text Fields", lead: "Text fields allow users to input, edit, and select text.", module: "@rmwc/textfield", styles: [
            '@material/textfield/dist/mdc.textfield.css',
            '@material/floating-label/dist/mdc.floating-label.css',
            '@material/notched-outline/dist/mdc.notched-outline.css',
            '@material/line-ripple/dist/mdc.line-ripple.css',
            '@material/ripple/dist/mdc.ripple.css',
            '@rmwc/icon/icon.css'
        ], docsLink: "https://material.io/develop/web/components/input-controls/text-field/", examples: generated_examples_json_1.default },
        react_1.default.createElement(_doc_utils_1.DocsSubtitle, null, "TextField Variants"),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Standard" },
            react_1.default.createElement(_1.TextField, { label: "standard..." })),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Outlined" },
            react_1.default.createElement(_1.TextField, { outlined: true, label: "outlined..." })),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Full Width" },
            react_1.default.createElement(_1.TextField, { fullwidth: true, placeholder: "fullWidth..." })),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "No Label" },
            react_1.default.createElement(_1.TextField, { placeholder: "No label" })),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Icons" },
            react_1.default.createElement(react_1.default.Fragment, null,
                react_1.default.createElement(_1.TextField, { icon: "search", trailingIcon: "close", label: "icon..." }),
                react_1.default.createElement(_1.TextField, { label: "trailingIcon...", trailingIcon: {
                        icon: 'close',
                        tabIndex: 0,
                        onClick: function () { return console.log('Clear'); }
                    } }))),
        react_1.default.createElement(_doc_utils_1.DocsSubtitle, null, "Textareas"),
        react_1.default.createElement(_doc_utils_1.DocsP, null, "You can make the TextField a textarea. Make sure to include `outlined` for proper styling You can optionally make help text always visible by passing an object as props with persistent set to true. Textareas can also have an optional character counter which will work with the maxLength property."),
        react_1.default.createElement(_doc_utils_1.DocsExample, null,
            react_1.default.createElement(_1.TextField, { textarea: true, outlined: true, fullwidth: true, label: "textarea...", rows: 8, maxLength: 20, characterCount: true, helpText: {
                    persistent: true,
                    validationMsg: true,
                    children: 'The field is required'
                } })),
        react_1.default.createElement(_doc_utils_1.DocsSubtitle, null, "Validation"),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Disabled" },
            react_1.default.createElement(_1.TextField, { disabled: true, label: "Disabled..." })),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Required" },
            react_1.default.createElement(_1.TextField, { required: true, label: "Required...", value: "" })),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Invalid" },
            react_1.default.createElement(_1.TextField, { invalid: true, label: "Invalid...", value: "#@!$", onChange: function () { } })),
        react_1.default.createElement(_doc_utils_1.DocsExample, { label: "Validation Pattern" },
            react_1.default.createElement(_1.TextField, { label: "Validate Pattern", pattern: "[A-Za-z]{3}" })),
        react_1.default.createElement(_doc_utils_1.DocsSubtitle, null, "HTML Input Types"),
        react_1.default.createElement(_doc_utils_1.DocsP, null, "A preview of how `material-components-web` handles styling input types for your browser."),
        react_1.default.createElement(_doc_utils_1.DocsExample, null,
            react_1.default.createElement(react_1.default.Fragment, null,
                react_1.default.createElement(_1.TextField, { label: "text", type: "text" }),
                react_1.default.createElement(_1.TextField, { label: "color", type: "color", style: { width: '6rem' } }),
                react_1.default.createElement(_1.TextField, { label: "date", type: "date" }),
                react_1.default.createElement(_1.TextField, { label: "datetime-local", type: "datetime-local" }),
                react_1.default.createElement(_1.TextField, { label: "month", type: "month" }),
                react_1.default.createElement(_1.TextField, { label: "range", type: "range" }),
                react_1.default.createElement(_1.TextField, { label: "time", type: "time" }),
                react_1.default.createElement(_1.TextField, { label: "week", type: "week" }))),
        react_1.default.createElement(_doc_utils_1.DocProps, { src: generated_props_json_1.default, components: [{ displayName: 'TextField', component: _1.TextField }] })));
}
exports.default = default_1;
exports.galleryExample = react_1.default.createElement(_1.TextField, { icon: "search", placeholder: "Search" });
