// @rmwc/textfield/dist/textfield-foundation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTextFieldFoundation = void 0;
var tslib_1 = require("tslib");
var react_1 = require("react");
var base_1 = require("@rmwc/base");
var textfield_1 = require("@material/textfield");
exports.useTextFieldFoundation = function (props) {
    var _a = tslib_1.__read(react_1.useState(false), 2), lineRippleActive = _a[0], setLineRippleActive = _a[1];
    var _b = tslib_1.__read(react_1.useState(0), 2), lineRippleCenter = _b[0], setLineRippleCenter = _b[1];
    var _c = tslib_1.__read(react_1.useState(), 2), notchWidth = _c[0], setNotchWidth = _c[1];
    var _d = tslib_1.__read(react_1.useState(false), 2), shakeLabel = _d[0], setShakeLabel = _d[1];
    var _e = tslib_1.__read(react_1.useState(false), 2), floatLabel = _e[0], setFloatlabel = _e[1];
    var characterCounter = react_1.useRef();
    var setCharacterCounter = function (api) {
        return (characterCounter.current = api);
    };
    var leadingIcon = react_1.useRef();
    var setLeadingIcon = function (api) {
        return (leadingIcon.current = api);
    };
    var trailingIcon = react_1.useRef();
    var setTrailingIcon = function (api) {
        return (trailingIcon.current = api);
    };
    var floatingLabel = react_1.useRef();
    var setFloatingLabel = function (api) {
        return (floatingLabel.current = api);
    };
    var _f = base_1.useFoundation({
        props: props,
        elements: { rootEl: true, inputEl: true },
        foundation: function (_a) {
            var rootEl = _a.rootEl, inputEl = _a.inputEl, getProps = _a.getProps;
            var getLabelAdapterMethods = function () {
                return {
                    shakeLabel: function (shouldShake) { return setShakeLabel(shouldShake); },
                    floatLabel: function (shouldFloat) {
                        var _a;
                        setFloatlabel((_a = getProps().floatLabel) !== null && _a !== void 0 ? _a : shouldFloat);
                    },
                    hasLabel: function () {
                        return !!getProps().label;
                    },
                    getLabelWidth: function () { var _a; return ((_a = floatingLabel.current) === null || _a === void 0 ? void 0 : _a.getWidth()) || 0; }
                };
            };
            var getLineRippleAdapterMethods = function () {
                return {
                    activateLineRipple: function () {
                        setLineRippleActive(true);
                    },
                    deactivateLineRipple: function () {
                        setLineRippleActive(false);
                    },
                    setLineRippleTransformOrigin: function (normalizedX) {
                        setLineRippleCenter(normalizedX);
                    }
                };
            };
            var getOutlineAdapterMethods = function () {
                return {
                    notchOutline: function (labelWidth) {
                        setNotchWidth(labelWidth);
                    },
                    closeOutline: function () {
                        var _a;
                        (_a = getProps().floatLabel) !== null && _a !== void 0 ? _a : setNotchWidth(undefined);
                    },
                    hasOutline: function () {
                        return !!getProps().outlined;
                    }
                };
            };
            var getInputAdapterMethods = function () {
                return {
                    registerInputInteractionHandler: function (evtType, handler) { return inputEl.addEventListener(evtType, handler); },
                    deregisterInputInteractionHandler: function (evtType, handler) { return inputEl.removeEventListener(evtType, handler); },
                    getNativeInput: function () { return inputEl.ref; }
                };
            };
            var getFoundationMap = function () {
                return {
                    characterCounter: characterCounter.current
                        ? characterCounter.current.getFoundation()
                        : undefined,
                    helperText: undefined,
                    leadingIcon: leadingIcon.current
                        ? leadingIcon.current.getFoundation()
                        : undefined,
                    trailingIcon: trailingIcon.current
                        ? trailingIcon.current.getFoundation()
                        : undefined
                };
            };
            return new textfield_1.MDCTextFieldFoundation(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign(tslib_1.__assign({ addClass: function (className) { return rootEl.addClass(className); }, removeClass: function (className) { return rootEl.removeClass(className); }, hasClass: function (className) { return rootEl.hasClass(className); }, registerTextFieldInteractionHandler: function (evtType, handler) { return rootEl.addEventListener(evtType, handler); }, deregisterTextFieldInteractionHandler: function (evtType, handler) { return rootEl.removeEventListener(evtType, handler); }, registerValidationAttributeChangeHandler: function (handler) {
                    var getAttributesList = function (mutationsList) {
                        return mutationsList.map(function (mutation) { return mutation.attributeName; });
                    };
                    if (inputEl.ref) {
                        var observer = new MutationObserver(function (mutationsList) {
                            return handler(getAttributesList(mutationsList));
                        });
                        var targetNode = inputEl.ref;
                        var config = { attributes: true };
                        targetNode && observer.observe(targetNode, config);
                        return observer;
                    }
                    return {};
                }, deregisterValidationAttributeChangeHandler: function (observer) {
                    observer && observer.disconnect();
                }, isFocused: function () {
                    return document.activeElement === inputEl.ref;
                } }, getInputAdapterMethods()), getLabelAdapterMethods()), getLineRippleAdapterMethods()), getOutlineAdapterMethods()), getFoundationMap());
        }
    }), foundation = _f.foundation, elements = tslib_1.__rest(_f, ["foundation"]);
    // Fixes bug #362
    // MDC breaks Reacts unidirectional data flow...
    // We have to capture the value before render
    // and then compare it to props.value after render in order to set
    // the appropriate foundation value without breaking its initial state
    var foundationValue = foundation.getValue();
    react_1.useEffect(function () {
        if (props.value !== undefined && props.value !== foundationValue) {
            foundation.setValue(String(props.value));
        }
    }, [props.value, foundation, foundationValue]);
    // Allow the user to float the label themselves
    react_1.useEffect(function () {
        if (props.floatLabel !== undefined) {
            foundation.notchOutline(props.floatLabel);
            // @ts-ignore unsafe adapter access
            foundation.adapter_.floatLabel(props.floatLabel);
        }
    }, [foundation, props.floatLabel]);
    return tslib_1.__assign({ shakeLabel: shakeLabel,
        floatLabel: floatLabel,
        notchWidth: notchWidth,
        lineRippleActive: lineRippleActive,
        lineRippleCenter: lineRippleCenter,
        setCharacterCounter: setCharacterCounter,
        setLeadingIcon: setLeadingIcon,
        setTrailingIcon: setTrailingIcon,
        setFloatingLabel: setFloatingLabel }, elements);
};
