// @rmwc/textfield/dist/textfield-character-count-foundation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTextFieldCharacterCountFoundation = void 0;
var tslib_1 = require("tslib");
var react_1 = require("react");
var base_1 = require("@rmwc/base");
var textfield_1 = require("@material/textfield");
exports.useTextFieldCharacterCountFoundation = function (props) {
    var _a = tslib_1.__read(react_1.useState(), 2), content = _a[0], setContent = _a[1];
    var _b = base_1.useFoundation({
        props: props,
        api: function (_a) {
            var foundation = _a.foundation;
            return {
                getFoundation: function () { return foundation; }
            };
        },
        elements: {},
        foundation: function () {
            return new textfield_1.MDCTextFieldCharacterCounterFoundation({
                setContent: function (content) {
                    setContent(content);
                }
            });
        }
    }), foundation = _b.foundation, elements = tslib_1.__rest(_b, ["foundation"]);
    return tslib_1.__assign({ content: content }, elements);
};
